smart-contracts/contracts/cl-vault/src/helpers.rs

```diff
@@ -290,139 +326,185 @@ pub fn get_max_utilization_for_ratio(
     }
 }
 
-pub fn get_liquidity_amount_for_unused_funds(
-    deps: DepsMut,
-    env: &Env,
-    additional_excluded_funds: (Uint128, Uint128),
-) -> Result<Decimal256, ContractError> {
-    // first get the ratio of token0:token1 in the position.
-    let p = get_position(deps.storage, &deps.querier)?;
-    // if there is no position, then we can assume that there are 0 unused funds
-    if p.position.is_none() {
-        return Ok(Decimal256::zero());
+// TODO, allocating funds gives a ratio for each position, we should theb
+pub fn allocate_funds_per_position(
+    _deps: DepsMut,
+    positions: Vec<(Position, FullPositionBreakdown)>,
+    spot_price: Decimal,
+    token0: Uint128,
+    token1: Uint128,
+) -> Result<Vec<(Position, Uint128, Uint128)>, ContractError> {
+    // get the wanted ratio per position
+    let ps = get_min_ratio_per_position(positions, spot_price)?;
+
+    let (total0, total1) = ps.iter().try_fold(
+        (Uint128::zero(), Uint128::zero()),
+        |(acc0, acc1), (_, pr)| -> Result<(Uint128, Uint128), ContractError> {
+            Ok((acc0.checked_add(pr.asset0)?, acc1.checked_add(pr.asset1)?))
+        },
+    )?;
+
+    // assuming we have the ratio wanted_token0/total_token0 and wanted_token1/total_token1
+    // ratio = min(wanted_token0/total_token0, wanted_token1/total_token1)
+    // for token0 and token1, we want to allocate
+    // allocate0 = token0 * ratio
+    // allocate1 = token1 * ratio
+    // afterwards we need to normalize all ratios to a 1 total
+    let ratios = ps.into_iter().map(|(ps, ps_ratio)| {
+        let r0 = Decimal::from_ratio(ps_ratio.asset0, total0);
+        let r1 = Decimal::from_ratio(ps_ratio.asset1, total1);
+        let ratio = min(r0, r1);
+
+        (ps, ratio)
+    });
+
+    let total_ratio = ratios
+        .clone()
+        .fold(Decimal::zero(), |acc, (_, ratio)| acc + ratio);
+
+    let psf: Result<Vec<(Position, Uint128, Uint128)>, ContractError> = ratios
+        .into_iter()
+        .map(|(ps, ratio)| {
+            let normalized = ratio / total_ratio;
+
+            Ok((ps, token0 * normalized, token1 * normalized))
+        })
+        .collect();
+    psf
+}
+
+/// per position, calculate the ratio of tokens that position needs, multiplied with the ratio of that position
+/// within the vault
+fn get_min_ratio_per_position(
+    positions: Vec<(Position, FullPositionBreakdown)>,
+    spot_price: Decimal,
+) -> Result<Vec<(Position, PositionRatio)>, ContractError> {
+    // per position, calculate the ratio of asset1 and asset2 a position needs
+
+    // for each position, we take some set amount of liquidity and calculate how many asset0 and asset1 those would give us
+    // we assume
+    let liquidity: Decimal256 = Decimal::new(100_000_000_000_000_000_000_000_u128.into()).into();
```

Comment:
Couldn't get the logic for using 100_000_000_000_000_000_000_000_u128 number here for liquidity.

Reply:
I'm using some big number so we don't have rounding errors, we want to take the same amount for each position to get some amount out, if we want we can lower it a bit anyway, but we want to be in atleast the 1000's to 100_000 tokens range imo